File: leetcode/src/alien-dictionary.py
from collections import deque

class Solution:
    def alienOrder(self, words: List[str]) -> str:
        letter_graph = dict()
        in_degree = dict()
        all_letters = set("".join(words))

        for letter in all_letters:
            letter_graph[letter] = set()
            in_degree[letter] = 0

        for word_idx in range(len(words)-1):
            word = words[word_idx]
            next_word = words[word_idx+1]

            for letter, next_letter in zip(word, next_word):
                if letter == next_letter:
                    continue
                if next_letter in letter_graph[letter]:
                    break
                letter_graph[letter].add(next_letter)
                in_degree[next_letter] += 1
                break
            else: #nobreak
                if len(next_word) < len(word):
                    return ""

        q = deque()
        topological_order = []
        for l, deg in in_degree.items():
            if deg == 0:
                q.append(l)

        while q:
            letter = q.popleft()
            topological_order.append(letter)
            for neighbour in letter_graph[letter]:
                in_degree[neighbour] -= 1
                if in_degree[neighbour] == 0:
                    q.append(neighbour)

        
        for _, deg in in_degree.items():
            if deg != 0:
                return ""

        return "".join(topological_order)

File: leetcode/src/alien-dictionary.cpp
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <string>

using namespace std;


class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_set<char> all_letters;
        for (const string& word : words) {
            all_letters.insert(word.begin(), word.end());
        }
        
        unordered_map<char, int> in_degree;
        unordered_map<char, unordered_set<char>> graph;

        for (char letter : all_letters) {
            in_degree[letter] = 0;
            unordered_set<char> s;
            graph.emplace(letter, s);
        } 

        for (int word_idx = 0; word_idx < words.size()-1; word_idx ++) {
            string word = words[word_idx];
            string next_word = words[word_idx + 1];
            int letter_idx;
            bool is_same = true;
            for (int letter_idx = 0; letter_idx < min(word.size(), next_word.size()); letter_idx ++) {
                char letter = word[letter_idx];
                char next_letter = next_word[letter_idx];
                if (letter == next_letter) {
                    continue;
                }
                is_same = false;
                if (graph[letter].find(next_letter) != graph[letter].end()) {
                    break;
                }
                graph[letter].insert(next_letter);
                in_degree[next_letter] ++;
                break;
            }
            if (is_same && word.size() > next_word.size()) {
                return "";
            }
        }

        queue<char> q;
        string topological_ordering;

        for (const auto item : in_degree) {
            if (item.second == 0) {
                q.push(item.first);
            }
        }

        while (!q.empty()) {
            char letter = q.front();
            q.pop();
            topological_ordering.push_back(letter);
            for (char neighbour : graph[letter]) {
                in_degree[neighbour] --;
                if (in_degree[neighbour] == 0) {
                    q.push(neighbour);
                }
            }
        }

        for (const auto item : in_degree) {
            if (item.second != 0) {
                return "";
            }
        }

        return topological_ordering;
    }
};